### ISSUE TITLE
Docker image for ARM64 (aarch64)

### ISSUE THREAD

**Issue by grubeli** [moigagoo/nimage#40]
[Docker image](https://hub.docker.com/r/nimlang/nim) is only for AMD64, in my current project I need arm64, is there a way to compile docker image for arm64?
Maybe having arm64 docker is also worthy?

**Comment by niv** [comment]
@moigagoo 

Here's my attempt: https://hub.docker.com/r/nivviv/nim

Dear reader: **_Please don't use my images. They're for personal/hobby projects and I make no guarantees._** If you want to steal the script to build your own images though, have at it!

As to getting arm64 into the official images:

You'll have to refactor the build system to use `buildx` instead, which is already part of Docker Desktop on mac, and should also be on Linux; then it can cross-compile locally (even on x64). Note, however, that buildx runs in qemu so depending on your CPU, this can take an hour or longer.

My images seem to be a bit smaller than yours. The main differentiator I had seen is combining the wget/unpack/build/rm into a single RUN layer. Dockerfile doesn't seem to be smart enough to combine them.

I haven't built ubuntu images (yet) because the alpine ones seem fine. I also haven't made ONBUILD variants because I think that's kind of obsolete with multi-FROM in docker. My repo readme has an example on how to use it. 

To reproduce this process in the official images:

* `buildx create --use --bootstrap` (run once or write a tester like I did so that it doesn't create lots of dummy envs)
* `docker buildx build ...... --platform linux/arm64,linux/amd64 --push .`  You need to say --push and it will  immediately push to dockerhub. buildx cannot import into the local registry for testing if you do multi-arch (but you can only give one --platform and say --load instead for testing; but then you can't push that same image).

**Comment by theAkito** [comment]
I was very surprised to notice the hard way, this project does not just naturally, self-explanatorily, use `buildx` to simply build multi-arch builds.

If this isn't already handled, I am more than willing to provide multi-arch builds to the related GitHub workflow...

**Comment by theAkito** [comment]
Hereby I kindly present you my multi-arch image.

https://github.com/theAkito/docker-nim

https://hub.docker.com/r/akito13/nim

It supports the following architectures.

```
linux/ppc64le
linux/arm/v7
linux/arm64/v8
linux/amd64
```

**Comment by moigagoo** [comment]
@theAkito I think you should chime in in this thread and propose your solution for the official Docker images: https://github.com/nim-lang/RFCs/issues/515

**Comment by theAkito** [comment]
> I really think the images in this repo didn't age well. First, they're still unofficial. Second, they're not automated to build in sync with new releases, I build and push new images manually.
> 
> If your images are better than these, please feel free to take over.

Sure, I'd be willing to take over!



> @theAkito I think you should chime in in this thread and propose your solution for the official Docker images: [nim-lang/RFCs#515](https://github.com/nim-lang/RFCs/issues/515)

Thanks for this, didn't know about this issue. Will comment now...

**Comment by moigagoo** [comment]
@theAkito

> Sure, I'd be willing to take over!

I've invited you as a contributor to this repo. I'll grant you full permissions to the repo after you accept the invitation.

**Comment by theAkito** [comment]
> I've invited you as a contributor to this repo. I'll grant you full permissions to the repo after you accept the invitation.

Thanks! Accepted the invitation. However, I think I would first wait for someone to respond at the RFC first, to get to an agreement what the next steps would be, before changing anything.

Am also not sure anymore, how I found this image, but isn't this image at least half official or is it totally unofficial & just randomy used by most Nim in Docker users? 🤔

**Comment by moigagoo** [comment]
@corv89 

> Hi, what's the status with arm64 nim containers?

What would be required to build those? I have a Mac with M3 processor, can I just build the images on it?

**Comment by corv89** [comment]
That would be ideal. I believe Github Actions also have Arm runners. Otherwise, Nim provides cross-compilation - I haven't looked into it yet however

Are your images built from CI, or do you build them manually? I wouldn't mind looking into Github Actions for you if that's missing

**Comment by theAkito** [comment]
> Thank you @theAkito , I'll try it with my project 🙏 

Sure, but the version is not up to date right now. If my image would be made the official one, I'd set up an automatic pipeline for it and I'd be willing to maintain the whole thing.

**Comment by corv89** [comment]
Your alpine image does build cleanly for me. I'd vouch for you, since this would complete arm64 support nicely in the ecosystem

**Comment by moigagoo** [comment]
> Sure, but the version is not up to date right now. If my image would be made the official one, I'd set up an automatic pipeline for it and I'd be willing to maintain the whole thing.

It's not for me to decide but I have nothing against it. I've been maintaining the images for the past several years manually and this process would benefit from being automated. I'll be happy to transfer the Docker Hub account to you.

Let's start a thread on the forum maybe?

**Comment by theAkito** [comment]
> > Sure, but the version is not up to date right now. If my image would be made the official one, I'd set up an automatic pipeline for it and I'd be willing to maintain the whole thing.
> 
> It's not for me to decide but I have nothing against it. I've been maintaining the images for the past several years manually and this process would benefit from being automated. I'll be happy to transfer the Docker Hub account to you.
> 
> Let's start a thread on the forum maybe?

Yes, sure, we can do that. I mean, most things, regarding the image, are already ready and when this is officially signed, I would have no problem in putting more time into the project and keeping it maintained. Everybody can check the repository and discuss open questions, which I can hopefully answer.

**Comment by theAkito** [comment]
@moigagoo https://forum.nim-lang.org/t/12741

**Comment by moigagoo** [comment]
@theAkito please contact me on moigagoo at duck dot com and send your Docker Hub username. I'll add you as an Admin to nimlang organization so that you could replace current images with yours.

**Comment by moigagoo** [comment]
@theAkito now you are an `owner` in the nimlang team and can do pretty much everything there.

**Comment by moigagoo** [comment]
The last thing I did with the images was I attempted to push the arm image from my Mac but accidentally replaced the amd64 ones pushed earlier from my Windows machine.

@theAkito please feel free to replace my images with yours. Please try to preserve backward compatibility: maintain the tags used earlier and keep the images for older Nim versions.

**Comment by moigagoo** [comment]
@theAkito if you want to utilize my Nim app that I've been using to build and push images, feel free to modify it to your needs. You are a `collaborator` in this repo and should be able to push your code directly.

**Comment by theAkito** [comment]
@moigagoo

Thanks for the tips.

I will see, how exactly I will implement the solution, once I get there. Not sure, when I will get to it, but I hope this week already.

**Comment by moigagoo** [comment]
OK it turns out supporting multiarch build is trivial so I've just added amd64, arm64, and arm builds to my build script.

Multiarch images for 2.2.2 are already available on Docker Hub.
